### PR TITLE
word2vec solver for blog post

### PR DESCRIPTION
**Pull request by joelgrus**

### CODE REVIEW

File: python/aristomini/common/wordtwovec.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    d1 = np.dot(v1, v1)
+    d2 = np.dot(v2, v2)
+
+    if d1 > 0.0 and d2 > 0.0:
+        return num / math.sqrt(d1 * d2)
+    else:
+        return 0.0
+
+
+class WordTwoVec:
```
Comment (aimichal) [inline python/aristomini/common/wordtwovec.py]: That's fun, can you make it less cute?

File: python/aristomini/common/wordtwovec.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    """a wrapper for gensim.Word2Vec with extra features we need"""
```
Comment (aimichal) [inline python/aristomini/common/wordtwovec.py]: What features? One more sentence will (probably) satisfy me.

File: python/aristomini/common/wordtwovec.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+        if vectors:
+            # if there are vector embeddings, take the vector average
+            return np.average(vectors, axis=0)
+        else:
+            # otherwise just return a zero vector
+            return np.zeros(self.model.vector_size)
+
+    def goodness(self, question_text: str, answer_text: str) -> float:
+        """how good is the answer for this question?"""
```
Comment (aimichal) [inline python/aristomini/common/wordtwovec.py]: The model in `MultipleChoiceQuestion.scala` is that an exam has choices, not answers. To match, this should take `question_stem` and `choice_text` strings.

File: python/aristomini/scripts/train_word2vec_model.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+from aristomini.common.wordtwovec import tokenizer
+
+from gensim.models import Word2Vec
+
+parser = argparse.ArgumentParser(description="train a word2vec model")
+parser.add_argument("input_sentences", help="file with the sentences to train on, one per line")
+parser.add_argument("output_model", help="where to save the model file")
+parser.add_argument("--size", type=int, default=50, help="dimension of the embedding")
+parser.add_argument("--window", type=int, default=5, help="size of the word window")
+parser.add_argument("--min_count", type=int, default=5,
```
Comment (aimichal) [inline python/aristomini/scripts/train_word2vec_model.py]: use hyphen: `--min-count`

File: python/aristomini/scripts/train_word2vec_model.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+"""
+
+import argparse
+
+from aristomini.common.wordtwovec import tokenizer
+
+from gensim.models import Word2Vec
+
+parser = argparse.ArgumentParser(description="train a word2vec model")
+parser.add_argument("input_sentences", help="file with the sentences to train on, one per line")
```
Comment (aimichal) [inline python/aristomini/scripts/train_word2vec_model.py]: Consider naming this `sentences_filename`

File: python/aristomini/common/wordtwovec.py (hunk 2 above)
Comment (joelgrus) [inline python/aristomini/common/wordtwovec.py]: done

File: python/aristomini/common/wordtwovec.py (hunk 3 above)
Comment (joelgrus) [inline python/aristomini/common/wordtwovec.py]: done

File: python/aristomini/scripts/train_word2vec_model.py (hunk 5 above)
Comment (joelgrus) [inline python/aristomini/scripts/train_word2vec_model.py]: I changed all the underscores to hyphens.

File: python/aristomini/scripts/train_word2vec_model.py (hunk 4 above)
Comment (joelgrus) [inline python/aristomini/scripts/train_word2vec_model.py]: see previous

File: python/aristomini/common/wordtwovec.py (hunk 1 above)
Comment (joelgrus) [inline python/aristomini/common/wordtwovec.py]: cute is good! (I'd rather leave it unless you feel *really really* strongly)

**Pull request merged**